File: apps/go/internal/chapter13/pca.go
// Package chapter13 は主成分分析（分散共分散行列・固有値分解・寄与率・射影）を扱う。
package chapter13

import (
	"fmt"
	"math"
	"slices"

	"gonum.org/v1/gonum/mat"
)

// Model は学習した主成分分析のモデル。
type Model struct {
	// Mean は列ごとの平均
	Mean []float64
	// Components は主成分を 1 行に 1 つずつ、寄与率の大きい順に並べたもの
	Components [][]float64
	// ExplainedVariance は主成分ごとの分散（固有値）
	ExplainedVariance []float64
	// ExplainedVarianceRatio は主成分ごとの寄与率
	ExplainedVarianceRatio []float64
}

// Loading は主成分の向きに対する 1 つの列の係数。
type Loading struct {
	// Column は列名
	Column string
	// Value は主成分の向きの成分（符号付き）
	Value float64
}

// validate は 1 件も無い、列の数がそろっていない、といった点数の並びを弾く。
func validate(x [][]float64) error {
	if len(x) == 0 {
		return fmt.Errorf("データが 1 件もありません")
	}

	for i, row := range x {
		if len(row) != len(x[0]) {
			return fmt.Errorf("%d 件目の列の数が違います: %d と %d", i+1, len(row), len(x[0]))
		}
	}

	if len(x[0]) == 0 {
		return fmt.Errorf("列が 1 つもありません")
	}

	return nil
}

// ColumnMeans は列ごとの平均を求める。
func ColumnMeans(x [][]float64) ([]float64, error) {
	if err := validate(x); err != nil {
		return nil, err
	}

	means := make([]float64, len(x[0]))

	for _, row := range x {
		for j, value := range row {
			means[j] += value
		}
	}

	for j := range means {
		means[j] /= float64(len(x))
	}

	return means, nil
}

// center は各列から平均を引く（中心化）。元のスライスは変更しない。
func center(x [][]float64, means []float64) [][]float64 {
	centered := make([][]float64, len(x))

	for i, row := range x {
		centered[i] = slices.Clone(row)
		for j := range centered[i] {
			centered[i][j] -= means[j]
		}
	}

	return centered
}

// CovarianceMatrix は列ごとの分散と、2 列ずつの共分散を並べた行列を返す。件数から 1 を引いた数で割る。
func CovarianceMatrix(x [][]float64) ([][]float64, error) {
	means, err := ColumnMeans(x)
	if err != nil {
		return nil, err
	}

	if len(x) < 2 {
		return nil, fmt.Errorf("分散共分散行列には 2 件以上が必要です: %d 件", len(x))
	}

	centered := center(x, means)
	columns := len(means)
	covariance := make([][]float64, columns)

	for j := range covariance {
		covariance[j] = make([]float64, columns)

		for k := range covariance[j] {
			sum := 0.0
			for _, row := range centered {
				sum += row[j] * row[k]
			}

			covariance[j][k] = sum / float64(len(x)-1)
		}
	}

	return covariance, nil
}

// Fit は分散共分散行列を固有値分解し、寄与率の大きい順に nComponents 個の主成分を求める。
// 固有値分解だけを gonum の mat.EigenSym に任せ、並べ替えと符号そろえは自分で行う。
func Fit(x [][]float64, nComponents int) (Model, error) {
	covariance, err := CovarianceMatrix(x)
	if err != nil {
		return Model{}, err
	}

	if nComponents < 1 || nComponents > len(covariance) {
		return Model{}, fmt.Errorf("主成分の数が範囲の外です: %d（列は %d）", nComponents, len(covariance))
	}

	values, vectors, err := eigen(covariance)
	if err != nil {
		return Model{}, err
	}

	total := 0.0
	for _, value := range values {
		total += value
	}

	if total == 0 {
		return Model{}, fmt.Errorf("すべての列の分散が 0 です")
	}

	means, err := ColumnMeans(x)
	if err != nil {
		return Model{}, err
	}

	model := Model{
		Mean:                   means,
		Components:             NormalizeSigns(vectors[:nComponents]),
		ExplainedVariance:      slices.Clone(values[:nComponents]),
		ExplainedVarianceRatio: make([]float64, nComponents),
	}

	for i, value := range model.ExplainedVariance {
		model.ExplainedVarianceRatio[i] = value / total
	}

	return model, nil
}

// eigen は対称行列を固有値分解し、固有値と固有ベクトル（1 行に 1 つ）を大きい順に返す。
// gonum の mat.EigenSym は小さい順に並べるので、ここで逆に並べ替える。
func eigen(symmetric [][]float64) ([]float64, [][]float64, error) {
	size := len(symmetric)
	flat := make([]float64, 0, size*size)

	for _, row := range symmetric {
		flat = append(flat, row...)
	}

	var decomposition mat.EigenSym
	if ok := decomposition.Factorize(mat.NewSymDense(size, flat), true); !ok {
		return nil, nil, fmt.Errorf("固有値分解に失敗しました")
	}

	ascending := decomposition.Values(nil)

	var vectors mat.Dense
	decomposition.VectorsTo(&vectors)

	values := make([]float64, size)
	descending := make([][]float64, size)

	for i := range size {
		source := size - 1 - i
		values[i] = ascending[source]
		descending[i] = mat.Col(nil, source, &vectors)
	}

	return values, descending, nil
}

// NormalizeSigns は、固有ベクトルは符号が逆でも同じ向きを表すので、
// 絶対値が最大の要素が正になるようにそろえる。元のスライスは変更しない。
func NormalizeSigns(components [][]float64) [][]float64 {
	normalized := make([][]float64, len(components))

	for i, component := range components {
		normalized[i] = slices.Clone(component)

		largest := 0.0
		for _, value := range component {
			if math.Abs(value) > math.Abs(largest) {
				largest = value
			}
		}

		if largest < 0 {
			for j := range normalized[i] {
				normalized[i][j] *= -1
			}
		}
	}

	return normalized
}

// Transform は平均を引いてから、データを主成分の向きに射影する。
func (m Model) Transform(x [][]float64) ([][]float64, error) {
	if err := validate(x); err != nil {
		return nil, err
	}

	if len(x[0]) != len(m.Mean) {
		return nil, fmt.Errorf("列の数が学習時と違います: %d と %d", len(x[0]), len(m.Mean))
	}

	centered := center(x, m.Mean)
	projected := make([][]float64, len(centered))

	for i, row := range centered {
		projected[i] = make([]float64, len(m.Components))

		for k, component := range m.Components {
			sum := 0.0
			for j, value := range row {
				sum += value * component[j]
			}

			projected[i][k] = sum
		}
	}

	return projected, nil
}

// ComponentsNeeded は累積寄与率がしきい値に届くまでの主成分の数を返す。
func ComponentsNeeded(ratios []float64, threshold float64) int {
	cumulative := 0.0

	for i, ratio := range ratios {
		cumulative += ratio
		if cumulative >= threshold {
			return i + 1
		}
	}

	return len(ratios)
}

// TopLoadings は主成分の係数の絶対値が大きい順に、列名と係数を k 個返す。
func TopLoadings(component []float64, columns []string, k int) ([]Loading, error) {
	if len(component) != len(columns) {
		return nil, fmt.Errorf("主成分の成分の数と列名の数が違います: %d と %d", len(component), len(columns))
	}

	loadings := make([]Loading, len(columns))
	for j, column := range columns {
		loadings[j] = Loading{Column: column, Value: component[j]}
	}

	slices.SortStableFunc(loadings, func(a, b Loading) int {
		return cmpDescending(math.Abs(a.Value), math.Abs(b.Value))
	})

	if k > len(loadings) {
		k = len(loadings)
	}

	return loadings[:k], nil
}

// cmpDescending は大きい順に並べるための比較。
func cmpDescending(a, b float64) int {
	switch {
	case a > b:
		return -1
	case a < b:
		return 1
	default:
		return 0
	}
}

File: apps/rust/src/chapter13/pca.rs
//! 主成分分析。分散共分散行列をヤコビ法で固有値分解し、寄与率の大きい順に主成分を並べる。

use ndarray::{Array1, Array2, Axis};

use super::{Error, Result};

/// 対角より外の成分が行列全体の大きさに対してこの割合より小さくなったら、
/// 固有値分解が終わったとみなす。絶対値で比べると、行列が大きいときに収束しない。
const TOLERANCE: f64 = 1e-12;
/// 1 回の掃き出しで対角より外の成分をすべて消す、その繰り返しの上限。
const MAX_SWEEPS: usize = 100;

/// 学習した主成分分析のモデル。
#[derive(Debug, Clone, PartialEq)]
pub struct PcaModel {
    /// 列ごとの平均。
    pub mean: Vec<f64>,
    /// 主成分を 1 行に 1 つずつ、寄与率の大きい順に並べた行列。
    pub components: Array2<f64>,
    /// 主成分ごとの分散（固有値）。
    pub explained_variance: Vec<f64>,
    /// 主成分ごとの寄与率。
    pub explained_variance_ratio: Vec<f64>,
}

/// 主成分の向きに対する 1 つの列の係数。
#[derive(Debug, Clone, PartialEq)]
pub struct Loading {
    pub column: String,
    pub value: f64,
}

/// 列ごとの平均。
pub fn column_means(x: &Array2<f64>) -> Vec<f64> {
    x.mean_axis(Axis(0))
        .map_or_else(Vec::new, |means| means.to_vec())
}

/// 各列から平均を引く（中心化）。
fn center(x: &Array2<f64>, means: &[f64]) -> Array2<f64> {
    let mut centered = x.clone();

    for mut row in centered.rows_mut() {
        for (value, mean) in row.iter_mut().zip(means) {
            *value -= mean;
        }
    }

    centered
}

/// 列ごとの分散と、2 列ずつの共分散を並べた行列。件数から 1 を引いた数で割る。
pub fn covariance_matrix(x: &Array2<f64>) -> Result<Array2<f64>> {
    if x.nrows() < 2 {
        return Err(Error::TooFewRows(x.nrows()));
    }

    let centered = center(x, &column_means(x));

    #[allow(clippy::cast_precision_loss)]
    let divisor = (x.nrows() - 1) as f64;

    Ok(centered.t().dot(&centered) / divisor)
}

/// 対称行列をヤコビ法で固有値分解する。固有ベクトルは列ごとに 1 つ返す。
///
/// ndarray 0.16 には固有値分解が無い（`ndarray-linalg` は LAPACK を要求する）ので、
/// 分散共分散行列に必要なぶん、つまり対称行列の場合だけを自作する。
pub fn jacobi_eigen(matrix: &Array2<f64>) -> Result<(Vec<f64>, Array2<f64>)> {
    let size = matrix.nrows();

    if matrix.ncols() != size {
        return Err(Error::NotSquare {
            rows: size,
            columns: matrix.ncols(),
        });
    }

    let mut work = matrix.clone();
    let mut vectors = Array2::eye(size);
    // 行列の大きさに合わせた「これ以上は消せない」しきい値
    let limit = TOLERANCE * frobenius_norm(matrix).max(f64::MIN_POSITIVE);

    for _ in 0..MAX_SWEEPS {
        if off_diagonal_norm(&work) < limit {
            return Ok((work.diag().to_vec(), vectors));
        }

        for p in 0..size {
            for q in (p + 1)..size {
                rotate(&mut work, &mut vectors, p, q, limit);
            }
        }
    }

    Err(Error::NotConverged(MAX_SWEEPS))
}

/// すべての成分の 2 乗和の平方根。
fn frobenius_norm(matrix: &Array2<f64>) -> f64 {
    matrix.iter().map(|value| value * value).sum::<f64>().sqrt()
}

/// 対角より外の成分の 2 乗和の平方根。
fn off_diagonal_norm(matrix: &Array2<f64>) -> f64 {
    let mut sum = 0.0;

    for ((i, j), value) in matrix.indexed_iter() {
        if i != j {
            sum += value * value;
        }
    }

    sum.sqrt()
}

/// (p, q) の成分が 0 になるように回転し、固有ベクトルにも同じ回転をかける。
fn rotate(work: &mut Array2<f64>, vectors: &mut Array2<f64>, p: usize, q: usize, limit: f64) {
    let apq = work[[p, q]];

    // すでに十分小さい成分は回さない。しきい値は収束の判定より細かくする
    if apq.abs() < limit * f64::EPSILON {
        return;
    }

    // tan(2θ) = 2 a_pq / (a_pp - a_qq) を解いて、cos と sin を求める
    let theta = 0.5 * (2.0 * apq).atan2(work[[p, p]] - work[[q, q]]);
    let (sin, cos) = theta.sin_cos();
    let size = work.nrows();

    for k in 0..size {
        let (kp, kq) = (work[[k, p]], work[[k, q]]);

        work[[k, p]] = cos * kp + sin * kq;
        work[[k, q]] = -sin * kp + cos * kq;
    }

    for k in 0..size {
        let (pk, qk) = (work[[p, k]], work[[q, k]]);

        work[[p, k]] = cos * pk + sin * qk;
        work[[q, k]] = -sin * pk + cos * qk;
    }

    for k in 0..size {
        let (kp, kq) = (vectors[[k, p]], vectors[[k, q]]);

        vectors[[k, p]] = cos * kp + sin * kq;
        vectors[[k, q]] = -sin * kp + cos * kq;
    }
}

/// 固有ベクトルは符号が逆でも同じ向きを表すので、絶対値が最大の要素が正になるようにそろえる。
pub fn normalize_signs(components: &Array2<f64>) -> Array2<f64> {
    let mut normalized = components.clone();

    for mut row in normalized.rows_mut() {
        let largest = row.iter().copied().fold(0.0_f64, |largest, value| {
            if value.abs() > largest.abs() {
                value
            } else {
                largest
            }
        });

        if largest < 0.0 {
            row.map_inplace(|value| *value = -*value);
        }
    }

    normalized
}

/// 分散共分散行列を固有値分解し、寄与率の大きい順に n_components 個の主成分を求める。
pub fn fit(x: &Array2<f64>, n_components: usize) -> Result<PcaModel> {
    let covariance = covariance_matrix(x)?;
    let (eigenvalues, eigenvectors) = jacobi_eigen(&covariance)?;

    if n_components == 0 || n_components > eigenvalues.len() {
        return Err(Error::ComponentCount {
            requested: n_components,
            columns: eigenvalues.len(),
        });
    }

    // 固有値の大きい順に並べ替える。値が同じときは元の順を保つ
    let mut order: Vec<usize> = (0..eigenvalues.len()).collect();
    order.sort_by(|left, right| eigenvalues[*right].total_cmp(&eigenvalues[*left]));

    let total: f64 = eigenvalues.iter().sum();
    let mut components = Array2::zeros((n_components, eigenvalues.len()));

    for (row, index) in order.iter().take(n_components).enumerate() {
        components.row_mut(row).assign(&eigenvectors.column(*index));
    }

    let explained_variance: Vec<f64> = order
        .iter()
        .take(n_components)
        .map(|index| eigenvalues[*index])
        .collect();

    Ok(PcaModel {
        mean: column_means(x),
        components: normalize_signs(&components),
        explained_variance_ratio: explained_variance
            .iter()
            .map(|variance| variance / total)
            .collect(),
        explained_variance,
    })
}

/// 平均を引いてから、データを主成分の向きに射影する。
pub fn transform(model: &PcaModel, x: &Array2<f64>) -> Array2<f64> {
    center(x, &model.mean).dot(&model.components.t())
}

/// 累積寄与率がしきい値に届くまでの主成分の数。
pub fn components_needed(ratios: &[f64], threshold: f64) -> usize {
    let mut cumulative = 0.0;

    for (index, ratio) in ratios.iter().enumerate() {
        cumulative += ratio;

        if cumulative >= threshold {
            return index + 1;
        }
    }

    ratios.len()
}

/// 主成分の係数の絶対値が大きい順に、列名と係数を k 個返す。
pub fn top_loadings(component: &Array1<f64>, columns: &[String], k: usize) -> Vec<Loading> {
    let mut loadings: Vec<Loading> = columns
        .iter()
        .zip(component)
        .map(|(column, value)| Loading {
            column: column.clone(),
            value: *value,
        })
        .collect();

    loadings.sort_by(|left, right| right.value.abs().total_cmp(&left.value.abs()));
    loadings.truncate(k);

    loadings
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::array;

    #[test]
    fn 分散共分散行列は件数から一を引いた数で割る() {
        // 1 列目の分散は ((-1)^2 + 0 + 1^2) / 2 = 1
        let x = array![[1.0, 2.0], [2.0, 4.0], [3.0, 6.0]];

        let covariance = covariance_matrix(&x).unwrap();

        assert!((covariance[[0, 0]] - 1.0).abs() < 1e-12);
        assert!((covariance[[0, 1]] - 2.0).abs() < 1e-12);
        assert!((covariance[[1, 1]] - 4.0).abs() < 1e-12);
    }

    #[test]
    fn データが一件だけなら分散を求められない() {
        assert_eq!(
            covariance_matrix(&array![[1.0, 2.0]])
                .unwrap_err()
                .to_string(),
            "主成分分析には 2 件以上のデータが必要です（1 件）"
        );
    }

    #[test]
    fn 対角行列の固有値はそのまま対角に並ぶ() {
        let (values, vectors) = jacobi_eigen(&array![[3.0, 0.0], [0.0, 1.0]]).unwrap();

        assert!((values[0] - 3.0).abs() < 1e-12);
        assert!((values[1] - 1.0).abs() < 1e-12);
        assert_eq!(vectors, Array2::<f64>::eye(2));
    }

    #[test]
    fn 完全に相関する二列の第一主成分は四十五度の向きになる() {
        // 2 列目が 1 列目と同じ値なので、ばらつきはすべて (1, 1) の向きにある
        let x = array![[1.0, 1.0], [2.0, 2.0], [3.0, 3.0], [4.0, 4.0]];

        let model = fit(&x, 2).unwrap();

        let root = 1.0 / 2.0_f64.sqrt();
        assert!((model.components[[0, 0]] - root).abs() < 1e-9, "{model:?}");
        assert!((model.components[[0, 1]] - root).abs() < 1e-9, "{model:?}");
        assert!((model.explained_variance_ratio[0] - 1.0).abs() < 1e-9);
        assert!(model.explained_variance_ratio[1].abs() < 1e-9);
    }

    #[test]
    fn 寄与率は大きい順に並ぶ() {
        // 1 列目のばらつきが 2 列目より大きい
        let x = array![[-2.0, -0.1], [-1.0, 0.1], [1.0, -0.1], [2.0, 0.1]];

        let model = fit(&x, 2).unwrap();

        assert!(model.explained_variance_ratio[0] > model.explained_variance_ratio[1]);
        assert!(
            (model.explained_variance_ratio.iter().sum::<f64>() - 1.0).abs() < 1e-12,
            "{model:?}"
        );
        assert!(model.components[[0, 0]].abs() > 0.99, "{model:?}");
    }

    #[test]
    fn 主成分の数は列の数までしか選べない() {
        let x = array![[1.0, 1.0], [2.0, 3.0]];

        assert!(fit(&x, 3).is_err());
        assert!(fit(&x, 0).is_err());
    }

    #[test]
    fn 符号は絶対値が最大の要素が正になるようにそろえる() {
        let normalized = normalize_signs(&array![[-0.8, 0.6], [0.8, -0.6]]);

        assert!((normalized[[0, 0]] - 0.8).abs() < 1e-12);
        assert!((normalized[[0, 1]] + 0.6).abs() < 1e-12);
        assert!((normalized[[1, 0]] - 0.8).abs() < 1e-12);
    }

    #[test]
    fn 射影した値は平均を引いてから主成分にかけた値になる() {
        let x = array![[1.0, 1.0], [2.0, 2.0], [3.0, 3.0]];
        let model = fit(&x, 1).unwrap();

        let projected = transform(&model, &x);

        // 平均 (2, 2) の点は原点に移る
        assert!(projected[[1, 0]].abs() < 1e-9, "{projected:?}");
        assert!(
            (projected[[0, 0]] + projected[[2, 0]]).abs() < 1e-9,
            "{projected:?}"
        );
    }

    #[test]
    fn 累積寄与率がしきい値に届くまでの数を返す() {
        let ratios = [0.5, 0.3, 0.2];

        assert_eq!(components_needed(&ratios, 0.4), 1);
        assert_eq!(components_needed(&ratios, 0.8), 2);
        assert_eq!(components_needed(&ratios, 0.9), 3);
        // しきい値に届かなければ、すべての主成分を使う
        assert_eq!(components_needed(&ratios, 1.5), 3);
    }

    #[test]
    fn 係数の絶対値が大きい順に列名を返す() {
        let columns = vec!["A".to_string(), "B".to_string(), "C".to_string()];

        let loadings = top_loadings(&array![0.1, -0.9, 0.5], &columns, 2);

        assert_eq!(loadings[0].column, "B");
        assert!((loadings[0].value + 0.9).abs() < 1e-12);
        assert_eq!(loadings[1].column, "C");
    }
}
